Adds dynamic client registration how-to guide

ddubson opened the pull request:
Closes #647

jgrandja commented on docs/modules/ROOT/pages/guides/how-to-dynamic-client-registration.adoc:
```diff
@@ -0,0 +1,174 @@
+[[sample.dcrAuthServerConfig]]
+[source,java]
+----
+include::{examples-dir}/main/java/sample/dcr/DcrConfiguration.java[]
```
We should provide a fully working configuration. Please take a look at 

https://github.com/spring-projects/spring-authorization-server/blob/e5fcbe422be1a73240ac7d8b40f165651fb01ab8/docs/src/main/java/sample/extgrant/SecurityConfig.java#L67

and how it makes use of `@fold:on` and `@fold:off` to only show the relevant configuration - but it can be expanded to show the complete configuration.

jgrandja commented on docs/src/main/java/sample/dcr/RegisteredClientConfiguration.java:
```diff
@@ -0,0 +1,32 @@
+				.scope("client.read") // <5>
+				.build();
+
+		JdbcRegisteredClientRepository registeredClientRepository = new JdbcRegisteredClientRepository(jdbcTemplate);
```
Use `InMemoryRegisteredClientRepository` instead to keep it simple

jgrandja commented on docs/src/main/java/sample/dcr/RegisteredClientConfiguration.java:
```diff
@@ -0,0 +1,32 @@
+	@Bean // <1>
+	public RegisteredClientRepository registeredClientRepository(JdbcTemplate jdbcTemplate) {
+		RegisteredClient initialClient = RegisteredClient.withId(UUID.randomUUID().toString())
+				.clientId("initial-client") // <2>
```
Let's come up with a better identifier, e.g. dcr-client, registrar-client, etc.

jgrandja commented on docs/modules/ROOT/pages/guides/how-to-dynamic-client-registration.adoc:
```diff
@@ -0,0 +1,174 @@
+
+[source,console]
+----
+curl -X POST --location "https://authserver.example.org/oauth2/token" --http1.1 \
```
Instead of using `curl`, let's keep the format generic, same as [Request the Access Token](https://docs.spring.io/spring-authorization-server/reference/guides/how-to-ext-grant-type.html#request-access-token)

jgrandja commented on docs/modules/ROOT/pages/guides/how-to-dynamic-client-registration.adoc:
```diff
@@ -0,0 +1,174 @@
+
+[source,console]
+----
+curl -X POST --location "https://authserver.example.org/connect/register" --http1.1 \
```
We should provide a fully working code sample. Since there currently is no client-side API for registering the client, we can still provide a programatic example using a `Map` of the claims to post via `WebClient` and receive a `Map` of the response.

jgrandja commented on docs/modules/ROOT/pages/guides/how-to-dynamic-client-registration.adoc:
```diff
@@ -0,0 +1,174 @@
+
+[source,console]
+----
+curl -X GET --location "<registration_client_uri>" \
```
We should provide a fully working code sample using `WebClient` and receive a `Map` of the response.

jgrandja reviewed (changes requested):
Thanks for the PR @ddubson. Please see review comments.

Also, please provide a test similar to [CustomCodeGrantTests](https://github.com/spring-projects/spring-authorization-server/blob/e5fcbe422be1a73240ac7d8b40f165651fb01ab8/docs/src/test/java/sample/extgrant/CustomCodeGrantTests.java#L38C27-L38C27)

ddubson commented:
Currently working on applying the suggested changes. Will notify when ready.

ddubson commented:
@jgrandja -- addressed your PR review comments. Ready for another review.

I've added an example dynamic client registration walkthrough, using `WebClient` as mentioned, but instead of using `Map`s, I opted for defining request and response classes. It makes it a bit easier to read, and easier to interact with `WebClient`.

In addition, I wanted to ensure that the sample walkthrough is fully tested, hence my leaving `assert`s on response objects within the walkthrough itself. I tried to balance it between readability and having the code be fully tested.

Let me know if this works or if modifications needed.

jgrandja commented:
Thanks for the updates @ddubson ! This is now merged.
